[user-streams] Add support functions for stream codegen

mlazos opened the pull request:
This PR adds utility functions for managing streams and a stream pool. 

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #174224
* #174223
* #165505
* #165391
* __->__ #165504
* #165390



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @chenyang78

mlazos committed:
[user-streams] Add support functions for stream codegen

[ghstack-poisoned]

mlazos committed:
Update on "[user-streams] Add support functions for stream codegen"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

[ghstack-poisoned]

mlazos committed:
Update on "[user-streams] Add support functions for stream codegen"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

[ghstack-poisoned]

mlazos committed:
Update on "[user-streams] Add support functions for stream codegen"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

[ghstack-poisoned]

mlazos committed:
Update on "[user-streams] Add support functions for stream codegen"

This PR adds utility functions for managing streams and a stream pool. 




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

[ghstack-poisoned]

eellison commented on torch/_inductor/stream_utils.py:
```diff
@@ -0,0 +1,138 @@
+    if stream_idx == 0:
+        return DEFAULT_STREAM
+    else:
+        return STREAM_NAME_TEMPLATE.format(stream_idx=stream_idx)
```
will this always be knowable? i think wed have to query what the stream is. especially bc streams are passed in as first class objects.

eellison reviewed (commented):
i don't think we need a stream pool, since pytorch eager manages its own. is it necessary ?

mlazos committed:
Update on "[user-streams] Add support functions for stream codegen"

This PR adds utility functions for managing streams and a stream pool. 




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

[ghstack-poisoned]

mlazos committed:
Update on "[user-streams] Add support functions for stream codegen"

This PR adds utility functions for managing streams and a stream pool. 




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

[ghstack-poisoned]

mlazos committed:
Update on "[user-streams] Add support functions for stream codegen"

This PR adds utility functions for managing streams and a stream pool. 




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

[ghstack-poisoned]

mlazos commented on torch/_inductor/stream_utils.py:
```diff
@@ -0,0 +1,138 @@
+                Defaults to 8.
+        """
+        self.pool_size: int = pool_size
+        self.stream_queue: queue.Queue[torch.cuda.Stream] = queue.Queue(maxsize=pool_size)
```
Yeah you're right, I'm curious why the NVIDIA folks did this, but I'll remove, creating streams is pretty lightweight as is.

mlazos committed:
Update on "[user-streams] Add support functions for stream codegen"

This PR adds utility functions for managing streams and a stream pool. 




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

[ghstack-poisoned]

mlazos committed:
Update on "[user-streams] Add support functions for stream codegen"

This PR adds utility functions for managing streams and a stream pool. 




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

[ghstack-poisoned]

mlazos commented on torch/_inductor/stream_utils.py:
```diff
@@ -0,0 +1,138 @@
+            stream (Optional[torch.cuda.Stream]): The CUDA stream to return to the pool.
+        """
+        if stream is not None:
+            self.stream_queue.put(stream)
```
I'm going to remove this

mlazos committed:
Update on "[user-streams] Add support functions for stream codegen"

This PR adds utility functions for managing streams and a stream pool. 




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

[ghstack-poisoned]

mlazos committed:
Update on "[user-streams] Add support functions for stream codegen"

This PR adds utility functions for managing streams and a stream pool. 




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

[ghstack-poisoned]
